File: Python/Visitor/Gemini 1.5 Pro/round1/code.py
from abc import ABC, abstractmethod

# Visitor interface
class ATMVisitor(ABC):

    @abstractmethod
    def visit_balance_inquiry(self, balance_inquiry):
        pass

    @abstractmethod
    def visit_cash_withdrawal(self, cash_withdrawal):
        pass

    @abstractmethod
    def visit_deposit(self, deposit):
        pass

    @abstractmethod
    def visit_pin_change(self, pin_change):
        pass

# Concrete visitor for handling transactions
class TransactionVisitor(ATMVisitor):

    def __init__(self, atm, account):
        self.atm = atm
        self.account = account

    def visit_balance_inquiry(self, balance_inquiry):
        print(f"Your balance is: ${self.account.balance:.2f}")

    def visit_cash_withdrawal(self, cash_withdrawal):
        amount = cash_withdrawal.amount
        if self.account.balance >= amount:
            self.account.balance -= amount
            self.atm.dispense_cash(amount)
            print(f"Withdrew ${amount:.2f}. New balance: ${self.account.balance:.2f}")
        else:
            print("Insufficient funds.")

    def visit_deposit(self, deposit):
        amount = deposit.amount
        self.account.balance += amount
        print(f"Deposited ${amount:.2f}. New balance: ${self.account.balance:.2f}")

    def visit_pin_change(self, pin_change):
        if self.account.verify_pin(pin_change.old_pin):
            self.account.change_pin(pin_change.new_pin)
            print("PIN changed successfully.")
        else:
            print("Incorrect old PIN.")

# ATM Elements (Visitable)
class ATMElement(ABC):

    @abstractmethod
    def accept(self, visitor: ATMVisitor):
        pass

class BalanceInquiry(ATMElement):
    def accept(self, visitor: ATMVisitor):
        visitor.visit_balance_inquiry(self)

class CashWithdrawal(ATMElement):
    def __init__(self, amount):
        self.amount = amount

    def accept(self, visitor: ATMVisitor):
        visitor.visit_cash_withdrawal(self)

class Deposit(ATMElement):
    def __init__(self, amount):
        self.amount = amount

    def accept(self, visitor: ATMVisitor):
        visitor.visit_deposit(self)

class PINChange(ATMElement):
    def __init__(self, old_pin, new_pin):
        self.old_pin = old_pin
        self.new_pin = new_pin

    def accept(self, visitor: ATMVisitor):
        visitor.visit_pin_change(self)

# Account class
class Account:
    def __init__(self, card_number, pin, balance=0):
        self.card_number = card_number
        self.pin = pin
        self.balance = balance
        self.attempts_left = 3

    def verify_pin(self, pin):
        if self.attempts_left > 0:
            if self.pin == pin:
                self.attempts_left = 3
                return True
            else:
                self.attempts_left -= 1
                print(f"Incorrect PIN. Attempts left: {self.attempts_left}")
                if self.attempts_left == 0:
                    print("Card retained due to too many incorrect PIN attempts.")
                return False
        else:
            print("Card is blocked. Please contact your bank.")
            return False

    def change_pin(self, new_pin):
        self.pin = new_pin

# ATM class
class ATM:
    def __init__(self):
        self.accounts = {}

    def add_account(self, account):
        self.accounts[account.card_number] = account

    def authenticate_user(self, card_number, pin):
        account = self.accounts.get(card_number)
        if account and account.verify_pin(pin):
            return account
        return None

    def dispense_cash(self, amount):
        # Logic for physically dispensing cash
        print(f"Dispensing ${amount:.2f} cash...")

    def run(self):
        card_number = input("Insert card (enter card number): ")
        pin = input("Enter PIN: ")
        account = self.authenticate_user(card_number, pin)
        if account:
            visitor = TransactionVisitor(self, account)
            while True:
                print("\nChoose an option:")
                print("1. Balance Inquiry")
                print("2. Cash Withdrawal")
                print("3. Deposit")
                print("4. PIN Change")
                print("5. Exit")
                choice = input("Enter choice: ")

                if choice == '1':
                    BalanceInquiry().accept(visitor)
                elif choice == '2':
                    amount = float(input("Enter amount to withdraw: "))
                    CashWithdrawal(amount).accept(visitor)
                elif choice == '3':
                    amount = float(input("Enter amount to deposit: "))
                    Deposit(amount).accept(visitor)
                elif choice == '4':
                    old_pin = input("Enter old PIN: ")
                    new_pin = input("Enter new PIN: ")
                    confirm_pin = input("Confirm new PIN: ")
                    if new_pin == confirm_pin:
                        PINChange(old_pin, new_pin).accept(visitor)
                    else:
                        print("New PINs do not match.")
                elif choice == '5':
                    print("Thank you for using the ATM!")
                    break
                else:
                    print("Invalid choice. Please try again.")
        else:
            print("Authentication failed.")


if __name__ == "__main__":
    atm = ATM()
    atm.add_account(Account("1234567890123456", "1234", 1000))
    atm.run()

File: Java/Visitor/Gemini 1.5 Pro/round1/code.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

// ATM class representing the context
public class ATM {
    private Map<String, Account> accounts;
    private Account currentAccount;
    private int maxAttempts;
    private Scanner input;

    public ATM() {
        accounts = new HashMap<>();
        accounts.put("1234567890123456", new Account("1234567890123456", "1234", 1000.0));
        maxAttempts = 3;
        input = new Scanner(System.in);
    }

    public static void main(String[] args) {
        ATM atm = new ATM();
        atm.start();
    }

    public void start() {
        System.out.println("Welcome to the ATM!");

        // Authentication
        if (authenticate()) {
            // User is authenticated
            System.out.println("Authentication successful.");
            showMenu();
        } else {
            // Authentication failed
            System.out.println("Authentication failed. Exiting...");
        }
    }

    private boolean authenticate() {
        System.out.print("Enter card number: ");
        String cardNumber = input.nextLine();

        if (accounts.containsKey(cardNumber)) {
            currentAccount = accounts.get(cardNumber);
            int attempts = 0;
            while (attempts < maxAttempts) {
                System.out.print("Enter PIN: ");
                String pin = input.nextLine();
                if (currentAccount.validatePin(pin)) {
                    return true;
                } else {
                    attempts++;
                    System.out.println("Incorrect PIN. Attempts remaining: " + (maxAttempts - attempts));
                }
            }
            System.out.println("Card retained due to exceeding maximum attempts.");
        } else {
            System.out.println("Invalid card number.");
        }
        return false;
    }

    private void showMenu() {
        ATMVisitor visitor = new ATMVisitorImpl(currentAccount, input);
        while (true) {
            System.out.println("\nSelect an option:");
            System.out.println("1. Check Balance");
            System.out.println("2. Withdraw Cash");
            System.out.println("3. Deposit Cash");
            System.out.println("4. Change PIN");
            System.out.println("5. Exit");

            int choice = input.nextInt();
            input.nextLine(); // consume newline

            switch (choice) {
                case 1:
                    visitor.visit(new BalanceInquiry());
                    break;
                case 2:
                    visitor.visit(new CashWithdrawal());
                    break;
                case 3:
                    visitor.visit(new CashDeposit());
                    break;
                case 4:
                    visitor.visit(new PinChange());
                    break;
                case 5:
                    System.out.println("Thank you for using the ATM. Goodbye!");
                    return;
                default:
                    System.out.println("Invalid option. Please try again.");
            }
        }
    }
}

// Account class
class Account {
    private String cardNumber;
    private String pin;
    private double balance;

    public Account(String cardNumber, String pin, double balance) {
        this.cardNumber = cardNumber;
        this.pin = pin;
        this.balance = balance;
    }

    public boolean validatePin(String pin) {
        return this.pin.equals(pin);
    }

    // Getters and setters
    public String getCardNumber() {
        return cardNumber;
    }

    public double getBalance() {
        return balance;
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public void setPin(String pin) {
        this.pin = pin;
    }
}

// Element interface in the Visitor pattern
interface ATMOperation {
    void accept(ATMVisitor visitor);
}

// Concrete element classes implementing ATMoperation
class BalanceInquiry implements ATMOperation {
    @Override
    public void accept(ATMVisitor visitor) {
        visitor.visit(this);
    }
}

class CashWithdrawal implements ATMOperation {
    @Override
    public void accept(ATMVisitor visitor) {
        visitor.visit(this);
    }
}

class CashDeposit implements ATMOperation {
    @Override
    public void accept(ATMVisitor visitor) {
        visitor.visit(this);
    }
}

class PinChange implements ATMOperation {
    @Override
    public void accept(ATMVisitor visitor) {
        visitor.visit(this);
    }
}

// Visitor interface
interface ATMVisitor {
    void visit(BalanceInquiry balanceInquiry);
    void visit(CashWithdrawal cashWithdrawal);
    void visit(CashDeposit cashDeposit);
    void visit(PinChange pinChange);
}

// Concrete visitor class implementing ATMVisitor
class ATMVisitorImpl implements ATMVisitor {
    private Account account;
    private Scanner input;

    public ATMVisitorImpl(Account account, Scanner input) {
        this.account = account;
        this.input = input;
    }

    @Override
    public void visit(BalanceInquiry balanceInquiry) {
        System.out.println("Your balance is: $" + account.getBalance());
    }

    @Override
    public void visit(CashWithdrawal cashWithdrawal) {
        System.out.print("Enter withdrawal amount: $");
        double amount = input.nextDouble();
        if (amount > 0 && amount <= account.getBalance()) {
            account.setBalance(account.getBalance() - amount);
            System.out.println("Withdrawal successful. Please take your cash.");
            System.out.println("Remaining balance: $" + account.getBalance());
        } else {
            System.out.println("Insufficient funds or invalid withdrawal amount.");
        }
    }

    @Override
    public void visit(CashDeposit cashDeposit) {
        System.out.print("Enter deposit amount: $");
        double amount = input.nextDouble();
        if (amount > 0) {
            account.setBalance(account.getBalance() + amount);
            System.out.println("Deposit successful.");
            System.out.println("New balance: $" + account.getBalance());
        } else {
            System.out.println("Invalid deposit amount.");
        }
    }

    @Override
    public void visit(PinChange pinChange) {
        System.out.print("Enter old PIN: ");
        String oldPin = input.nextLine();
        if (account.validatePin(oldPin)) {
            System.out.print("Enter new PIN: ");
            String newPin1 = input.nextLine();
            System.out.print("Confirm new PIN: ");
            String newPin2 = input.nextLine();
            if (newPin1.equals(newPin2)) {
                account.setPin(newPin1);
                System.out.println("PIN changed successfully.");
            } else {
                System.out.println("New PINs do not match.");
            }
        } else {
            System.out.println("Incorrect old PIN.");
        }
    }
}